### PR TITLE
Disallowing udp for lb rules for haproxy

### PR DESCRIPTION
**Pull request by davidjumani**
### Description

Since CloudStack uses HAproxy as a load balancer which doesn't support udp, adding checks to prevent udp from being passed as the protocol
https://github.com/haproxy/haproxy/issues/62

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### How Has This Been Tested?

```
(localhost) 🐱 > create loadbalancerrule publicipid=afe66a42-2ebc-476b-bf5f-8954fcc42f32 protocol=udp publicport=80 privateport=80 algorithm=roundrobin name=test
🙈 Error: (HTTP 431, error code 9999) LB service provider cannot support this rule
```

### CODE REVIEW

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
code lgtm

**Comment by ravening** [comment]
@davidjumani it does support udp

```
(local) 🐱 > create loadbalancerrule name=test privateport=80 publicport=8080 protocol=udp account=admin domainid=04dfd1c6-c28e-11ea-9a43-0617bc003377 networkid=d396de4e-0e23-4e3b-a19a-1fe1ac35227c publicipid=6275b645-fac5-4b03-a2f1-40848655ffd0 algorithm=roundrobin
{
  "loadbalancer": {
    "account": "admin",
    "algorithm": "roundrobin",
    "cidrlist": "",
    "domain": "ROOT",
    "domainid": "04dfd1c6-c28e-11ea-9a43-0617bc003377",
    "fordisplay": true,
    "id": "d1883751-9e55-4104-a522-c6c5c09a123b",
    "name": "test",
    "networkid": "d396de4e-0e23-4e3b-a19a-1fe1ac35227c",
    "privateport": "80",
    "protocol": "udp",
    "publicip": "10.10.20.1",
    "publicipid": "6275b645-fac5-4b03-a2f1-40848655ffd0",
    "publicport": "8080",
    "state": "Add",
    "tags": [],
    "zoneid": "8ae93aed-7db1-4a1a-9030-db0efac6ba62",
    "zonename": "mgt122-10"
  }
}
```

**Comment by davidjumani** [comment]
@ravening CloudStack allows a udp lb rule to be created but haproxy itself which is internally used for load balancing doesn't support udp load balancing
Also the cloudstack treats all lb rules as tcp
https://github.com/apache/cloudstack/blob/master/api/src/main/java/org/apache/cloudstack/api/command/user/loadbalancer/CreateLoadBalancerRuleCmd.java#L340
https://github.com/apache/cloudstack/blob/master/core/src/main/java/com/cloud/network/HAProxyConfigurator.java#L48

**Comment by ravening** [comment]
> @ravening CloudStack allows a udp lb rule to be created but haproxy itself which is internally used for load balancing doesn't support udp load balancing
> Also cloudstack treats all lb rules as tcp
> https://github.com/apache/cloudstack/blob/master/api/src/main/java/org/apache/cloudstack/api/command/user/loadbalancer/CreateLoadBalancerRuleCmd.java#L340
> https://github.com/apache/cloudstack/blob/master/core/src/main/java/com/cloud/network/HAProxyConfigurator.java#L48

@davidjumani 
yeah checked the haproxy docs and it doesnt support udp.
will there be a possibility where we can have different type of loadbalancers with udp support?

**Comment by davidjumani** [comment]
> @davidjumani
> yeah checked the haproxy docs and it doesnt support udp.
> will there be a possibility where we can have different type of loadbalancers with udp support?

@ravening There are options such as nginx which can be explored later on if there's a need for udp load balancing (since no one noticed that it didn't work so far), so raising this PR so users won't be misled

**Comment by DaanHoogland** [comment]
let's link to #4481 here, so we have context when discussing implementation.

**Comment by ravening** [comment]
@davidjumani @DaanHoogland pr for haproxy support is here https://github.com/apache/cloudstack/pull/4141

**Comment by DaanHoogland** [comment]
@ravening those are extra features for configuring/fine tuning the loadbalancer as it exists for tcp. the issue I linked is about missing support for udp and conflicts between udp and tcp definitions coexisting. (related but still unrelated)

**Comment by rhtyd** [comment]
@weizhouapache since you've worked a lot of haproxy/lb, do you agree with this PR?

File: api/src/main/java/org/apache/cloudstack/api/command/user/loadbalancer/CreateLoadBalancerRuleCmd.java [hunk 1]
```diff
@@ -255,6 +255,12 @@ public Boolean getOpenFirewall() {
     }
 
     public String getLbProtocol() {
+        if (lbProtocol != null) {
+            lbProtocol = lbProtocol.toLowerCase();
+            if (!lbProtocol.equals(NetUtils.TCP_PROTO) && !lbProtocol.equals(NetUtils.TCP_PROXY_PROTO)) {
+                throw new InvalidParameterValueException("Invalid protocool. Supported protocols are tcp and tcp-proxy");
```
Comment (weizhouapache) [inline api/src/main/java/org/apache/cloudstack/api/command/user/loadbalancer/CreateLoadBalancerRuleCmd.java:261]: @davidjumani as far as I remember, 'ssl' is also a valid protocol

File: api/src/main/java/org/apache/cloudstack/api/command/user/loadbalancer/CreateLoadBalancerRuleCmd.java [hunk 2]
```diff
@@ -114,7 +114,7 @@
         + "rule will be created for. Required when public Ip address is not associated with any Guest network yet (VPC case)")
     private Long networkId;
 
-    @Parameter(name = ApiConstants.PROTOCOL, type = CommandType.STRING, description = "The protocol for the LB such as tcp, udp or tcp-proxy.")
+    @Parameter(name = ApiConstants.PROTOCOL, type = CommandType.STRING, description = "The protocol for the LB such as tcp or tcp-proxy.")
     private String lbProtocol;
```
Comment (weizhouapache) [inline api/src/main/java/org/apache/cloudstack/api/command/user/loadbalancer/CreateLoadBalancerRuleCmd.java:118]: @davidjumani might 'udp' be supportd by some lb hardware devices ?

**Comment by weizhouapache** [comment]
@davidjumani fyi,
In my pr for load balancer configurations #4141, there is a check on haproxy rule

https://github.com/apache/cloudstack/pull/4141/files#diff-1fea20ffdf01a98062ef48dfd3afa97d6fb0355f2f8c84acddc5115876f0188dR792-R806

line 792 to 806 in server/src/main/java/com/cloud/network/router/NetworkHelperImpl.java

**Comment by rhtyd** [comment]
That's a good point, some LB providers may actually support UDP but in most real world usecases would LB over tcp.

File: api/src/main/java/org/apache/cloudstack/api/command/user/loadbalancer/CreateLoadBalancerRuleCmd.java (hunk 2 above)
Comment (davidjumani) [reply api/src/main/java/org/apache/cloudstack/api/command/user/loadbalancer/CreateLoadBalancerRuleCmd.java:118]: @weizhouapache good point, I'll move the check to the NetworkHelperImpl.java

**Comment by davidjumani** [comment]
@weizhouapache Added the check under `validateHAProxyLBRule`

**Comment by weizhouapache** [comment]
> @weizhouapache Added the check under `validateHAProxyLBRule`

@davidjumani I will test it

**Review by weizhouapache — APPROVED** [review]
looks good.

create a udp lb rule, got exception below
```
2020-11-27 07:41:12,538 DEBUG [c.c.n.r.NetworkHelperImpl] (qtp1430439149-16:ctx-b16df490 ctx-85efa4e7) (logid:0c5425d3) Can't create LB rule as haproxy does not support udp
2020-11-27 07:41:12,539 WARN  [c.c.n.l.LoadBalancingRulesManagerImpl] (qtp1430439149-16:ctx-b16df490 ctx-85efa4e7) (logid:0c5425d3) Failed to create load balancer due to
com.cloud.exception.InvalidParameterValueException: LB service provider cannot support this rule
```

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Review by rhtyd — APPROVED** [review]

**Pull request merged**